### ISSUE TITLE
AltGr modifier does not allow "Modifier when held" option.

### ISSUE THREAD

**Issue by skelendal** [Dygmalab/Bazecor#160]
**Is your feature request related to a problem? Please describe.**
"Modifier when held, normal key otherwise" option is not available for AltGr modifier. Not necessarily a big deal but would be very nice and it would be consistent with the other modifiers.

**Comment by michaelmejaeger** [comment]
This is a bug in my opinion: when checking the "AltGr"-Checkbox the Switch "Modifier when held, normal key otherwise" is inactive. This works with all other options (like Control, Shift, etc.). It would be great if also AltGr could be used for the Mod-Tap function.

**Comment by michix** [comment]
This is a bug in my opinion: when checking the "AltGr"-Checkbox the Switch "Modifier when held, normal key otherwise" is inactive. This works with all other options (like Control, Shift, etc.). It would be great if also AltGr could be used for the Mod-Tap function.

**Comment by skelendal** [comment]
Yes, this should be labeled as a bug instead of an enchancement but I can't change it.

**Comment by AlexDygma** [comment]
Hey @skelendal, @michaelmejaeger!

I moved it to bug

but the root cause of this is that there is not enough address space for the altGr combination to have a "when held" action on the firmware.

we will have to change the whole memory map to address this, so in the mean time, i am inclined to grey out the options when altGr is selected

**Comment by jeancroy** [comment]
> there is not enough address space

In theory you can already store any combination of modifier + key
Then you need a single flag to morph "modifier(s) + key" into "modifer(s) when held, key when pressed"

It's just not made that way. And you're right it's a large change. But could be needed to add more layer or macro. And it would be a shame to loose compatibility with base firmware.

**Comment by AlexDygma** [comment]
Hey @jeancroy !

Ofc if we change enough the way we codify the flags we can make enough room for this, but the reason it's implemented as it is is to allow the firmware to add all the modifiers to the same key as each one of them is capable of being flagged to true at the same time.

you can see the way the keys are defined here:
https://github.com/Dygmalab/Kaleidoscope/blob/7bac53de106c42ffda889e6854abc06cf43a3c6f/src/kaleidoscope/key_defs.h#L157

As you will see, there is a mask that's used bit by bit to allow different combinations for the pressed key, and i'ts already full, so if we wanted to add the right-alt + something else, it doesn't fit. we would have to double the mask's bits to make room or mux the bits so we can better make use of it.

on either case this change requires to modify the firmware very agressively or make a "demuxer" that's able to treat the mask before being treated by the rest of the FW.

**Comment by skelendal** [comment]
@AlexDygma I'm really not an expert in this and therefore this might be a totally incorrect assumption but I couldn't find mask _B00100000_ in the definition so wouldn't that mean it's available?

**Comment by AlexDygma** [comment]
> @AlexDygma I'm really not an expert in this and therefore this might be a totally incorrect assumption but I couldn't find mask _B00100000_ in the definition so wouldn't that mean it's available?

Yep, it's not on the file, but if you search for it on the code, in reality it's used as the macro flag!

so unless we don't want macros, it's taken 😞

**Comment by jeancroy** [comment]
@AlexDygma What's your plan on tapdance ? "Superpower".
It looks like those will need to be way more expressive...
And probably superset current dual use in functionality.

**Comment by AlexDygma** [comment]
> @AlexDygma What's your plan on tapdance ? "Superpower".
> It looks like those will need to be way more expressive...
> And probably superset current dual use in functionality.

TapDance is the first plugin that we are going to enable next, but we need to integrate it as smoothly as we can in Bazecor, and that doesn't seem easy.

We have a few plans to rework the whole interface that will facilitate a lot the configuration of this type of specific modifiers.
